### ISSUE TITLE
Ganesha crash at nfs4_Fattr_Free()->gsh_free()

### ISSUE THREAD

**Issue by skmprabhu252** [nfs-ganesha/nfs-ganesha#1085]
Ganesha crashed during below testcase
             
              • Thread-1 -> Create 10 files in a loop. (while(true); do for i in {1..10} ; do touch file$i; done; done)
              • Thread-2 -> Delete created files in the loop. (while(true); do for i in {1..10} ; do rm -f file$i; done; done
              • Thread-3 -> Try to take a file lock using fcntl()
             

As per the below logs, in frame 8, the variable p points to a valid address.  This looks like a heap memory is corrupted. 

(gdb) where
#0  0x00007f602e41db8f in raise () from /lib64/libpthread.so.0
#1  0x00007f603010edd1 in crash_handler (signo=6, info=0x7f602852ce70, ctx=0x7f602852cd40) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/MainNFSD/nfs_init.c:256
#2  <signal handler called>
#3  0x00007f602dc72acf in raise () from /lib64/libc.so.6
#4  0x00007f602dc45ea5 in abort () from /lib64/libc.so.6
#5  0x00007f602dcb3cd7 in __libc_message () from /lib64/libc.so.6
#6  0x00007f602dcbafdc in malloc_printerr () from /lib64/libc.so.6
#7  0x00007f602dcbcb64 in _int_free () from /lib64/libc.so.6
#8  0x00007f60301ed551 in gsh_free (p=0x7f6000047b20) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/include/abstract_mem.h:257
#9  0x00007f60301f2a07 in nfs4_Fattr_Free (fattr=0x7f600c063770) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs_proto_tools.c:3516
#10 0x00007f60301c6afa in nfs4_op_getattr_Free (res=0x7f600c063760) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs4_op_getattr.c:241
#11 0x00007f60301beb96 in nfs4_Compound_FreeOne (res=0x7f600c063760) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs4_Compound.c:1418
#12 0x00007f60301bed73 in release_nfs4_res_compound (res_compound4_ex=0x7f600c061d70) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs4_Compound.c:1447
#13 0x00007f60301bede7 in nfs4_Compound_Free (res=0x7f600c066bb0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs4_Compound.c:1471
#14 0x00007f6030142a52 in nfs_dupreq_free_dupreq (dv=0x7f600c043ce0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/RPCAL/nfs_dupreq.c:913
#15 0x00007f6030142c04 in dupreq_entry_put (dv=0x7f600c043ce0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/RPCAL/nfs_dupreq.c:946
#16 0x00007f6030144b34 in nfs_dupreq_finish (reqnfs=0x7f60000383b0, rc=NFS_REQ_OK) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/RPCAL/nfs_dupreq.c:1318
#17 0x00007f603010762a in complete_request (reqdata=0x7f60000383b0, rc=NFS_REQ_OK) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/MainNFSD/nfs_worker_thread.c:809
#18 0x00007f6030109b67 in nfs_rpc_process_request (reqdata=0x7f60000383b0, retry=false) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/MainNFSD/nfs_worker_thread.c:1538
#19 0x00007f603010a0bf in nfs_rpc_valid_NFS (req=0x7f60000383b0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/MainNFSD/nfs_worker_thread.c:1732
#20 0x00007f602fe82741 in svc_vc_decode (req=0x7f60000383b0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/libntirpc/src/svc_vc.c:1046
#21 0x00007f602fe7dc5f in svc_request (xprt=0x7f601c000da0, xdrs=0x7f600002fa80) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/libntirpc/src/svc_rqst.c:1202
#22 0x00007f602fe82646 in svc_vc_recv (xprt=0x7f601c000da0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/libntirpc/src/svc_vc.c:1019
#23 0x00007f602fe7dbdf in svc_rqst_xprt_task_recv (wpe=0x7f601c001068) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/libntirpc/src/svc_rqst.c:1183
#24 0x00007f602fe7e773 in svc_rqst_epoll_loop (wpe=0x5343c98) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/libntirpc/src/svc_rqst.c:1564
#25 0x00007f602fe89c3a in work_pool_thread (arg=0x7f6008003fb0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/libntirpc/src/work_pool.c:183
#26 0x00007f602e4131ca in start_thread () from /lib64/libpthread.so.0
#27 0x00007f602dc5de73 in clone () from /lib64/libc.so.6

(gdb) f 8
#8  0x00007f60301ed551 in gsh_free (p=0x7f6000047b20) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/include/abstract_mem.h:257
257 free(p);
(gdb) p p
$1 = (void *) 0x7f6000047b20
(gdb) p *p
Attempt to dereference a generic pointer.
(gdb) list
252 * [@param](https://jazz07.rchland.ibm.com:21443/jazz/users/param)[in] p Block of memory to free.
253 */
254 static inline void
255 gsh_free(void *p)
256 {
257 free(p);
258 }
259
260 /**
261 * [@brief](https://jazz07.rchland.ibm.com:21443/jazz/users/brief) Free a block of memory with size

(gdb) f 9
#9  0x00007f60301f2a07 in nfs4_Fattr_Free (fattr=0x7f600c063770) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs_proto_tools.c:3516
3516 gsh_free(fattr->attr_vals.attrlist4_val);
(gdb) list
3511 }
3512
3513 void nfs4_Fattr_Free(fattr4 *fattr)
3514 {
3515 if (fattr->attr_vals.attrlist4_val != NULL) {
3516 gsh_free(fattr->attr_vals.attrlist4_val);
3517 fattr->attr_vals.attrlist4_val = NULL;
3518 fattr->attr_vals.attrlist4_len = 0;
3519 }
3520 }
(gdb) p fattr->attr_vals.attrlist4_val
$2 = 0x7f6000047b20 ""
(gdb) p fattr->attr_vals
$4 = {attrlist4_len = 160, attrlist4_val = 0x7f6000047b20 ""}

(gdb) f 10
#10 0x00007f60301c6afa in nfs4_op_getattr_Free (res=0x7f600c063760) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs4_op_getattr.c:241
241 nfs4_Fattr_Free(&resp->GETATTR4res_u.resok4.obj_attributes);
(gdb) list
236 void nfs4_op_getattr_Free(nfs_resop4 *res)
237 {
238 GETATTR4res *resp = &res->nfs_resop4_u.opgetattr;
239
240 if (resp->status == NFS4_OK)
241 nfs4_Fattr_Free(&resp->GETATTR4res_u.resok4.obj_attributes);
242 } /* nfs4_op_getattr_Free */

(gdb) p res->nfs_resop4_u.opgetattr
$5 = {status = NFS4_OK, GETATTR4res_u = {resok4 = {obj_attributes = {attrmask = {bitmap4_len = 2, map = {1048858, 11575866, 0}}, attr_vals = {attrlist4_len = 160, attrlist4_val = 0x7f6000047b20 ""}}}}}

(gdb) f 11
#11 0x00007f60301beb96 in nfs4_Compound_FreeOne (res=0x7f600c063760) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs4_Compound.c:1418
1418 optabv4[opcode].free_res(res);

(gdb) list
1413 {
1414 int opcode;
1415
1416 opcode = (res->resop != NFS4_OP_ILLEGAL)
1417     ? res->resop : 0; /* opcode 0 for illegals */
1418 optabv4[opcode].free_res(res);
1419 }
1420
(gdb) p res->nfs_resop4_u.opgetattr
$9 = {status = NFS4_OK, GETATTR4res_u = {resok4 = {obj_attributes = {attrmask = {bitmap4_len = 2, map = {1048858, 11575866, 0}}, attr_vals = {attrlist4_len = 160, attrlist4_val = 0x7f6000047b20 ""}}}}}
(gdb) p res->resop
$10 = NFS4_OP_GETATTR

(gdb) f 12
#12 0x00007f60301bed73 in release_nfs4_res_compound (res_compound4_ex=0x7f600c061d70) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs4_Compound.c:1447
1447 nfs4_Compound_FreeOne(val);
(gdb) list
1442
1443 if (val) {
1444 /* !val is an error case, but it can occur, so avoid
1445 * indirect on NULL
1446 */
1447 nfs4_Compound_FreeOne(val);
1448 }
1449 }
1450
1451 gsh_free(res_compound4->resarray.resarray_val);
(gdb) p val
$11 = (nfs_resop4 *) 0x7f600c063760
(gdb) p val.nfs_resop4_u.opgetattr
$13 = {status = NFS4_OK, GETATTR4res_u = {resok4 = {obj_attributes = {attrmask = {bitmap4_len = 2, map = {1048858, 11575866, 0}}, attr_vals = {attrlist4_len = 160, attrlist4_val = 0x7f6000047b20 ""}}}}}

   1421 void release_nfs4_res_compound(struct COMPOUND4res_extended *res_compound4_ex)
   1422 {
   1423         unsigned int i = 0;
   1424         int32_t refcnt = atomic_dec_int32_t(&res_compound4_ex->res_refcnt);
   1425         struct COMPOUND4res *res_compound4 = &res_compound4_ex->res_compound4;
   1426
   1427         assert(refcnt >= 0);
   1428
   1429         if (refcnt > 0) {
   1430                 LogFullDebugAlt(COMPONENT_NFS_V4, COMPONENT_SESSIONS,
   1431                              "Skipping free of NFS4 result %p refcnt %"PRIi32,
   1432                              res_compound4_ex, refcnt);
   1433                 return;
   1434         }
   1435
   1436         LogFullDebugAlt(COMPONENT_NFS_V4, COMPONENT_SESSIONS,
   1437                      "Compound Free %p (resarraylen=%i)",
   1438                      res_compound4_ex, res_compound4->resarray.resarray_len);
   1439
   1440         for (i = 0; i < res_compound4->resarray.resarray_len; i++) {
   1441                 nfs_resop4 *val = &res_compound4->resarray.resarray_val[i];
   1442
   1443                 if (val) {
   1444                         /* !val is an error case, but it can occur, so avoid
   1445                          * indirect on NULL
   1446                          */
   1447                         nfs4_Compound_FreeOne(val);
   1448                 }
   1449         }

(gdb) p res_compound4->resarray.resarray_val[0].nfs_resop4_u.opgetattr
$17 = {status = NFS4_OK, GETATTR4res_u = {resok4 = {obj_attributes = {attrmask = {bitmap4_len = 0, map = {0, 0, 0}}, attr_vals = {attrlist4_len = 0, attrlist4_val = 0x0}}}}}
(gdb) p res_compound4->resarray.resarray_val[1].nfs_resop4_u.opgetattr
$18 = {status = NFS4_OK, GETATTR4res_u = {resok4 = {obj_attributes = {attrmask = {bitmap4_len = 31, map = {0, 0, 0}}, attr_vals = {attrlist4_len = 0, attrlist4_val = 0x0}}}}}
(gdb) p res_compound4->resarray.resarray_val[2].nfs_resop4_u.opgetattr
$19 = {status = NFS4_OK, GETATTR4res_u = {resok4 = {obj_attributes = {attrmask = {bitmap4_len = 2, map = {1048858, 11575866, 0}}, attr_vals = {attrlist4_len = 160, attrlist4_val = 0x7f6000047b20 ""}}}}}
(gdb)

(gdb) p res_compound4_ex->res_compound4
$22 = {status = NFS4_OK, tag = {utf8string_len = 0, utf8string_val = 0x0}, resarray = {resarray_len = 3, resarray_val = 0x7f600c063620}}

(gdb) f 13
#13 0x00007f60301bede7 in nfs4_Compound_Free (res=0x7f600c066bb0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/Protocols/NFS/nfs4_Compound.c:1471
1471 release_nfs4_res_compound(res->res_compound4_extended);
(gdb) list
1466 * [@param](https://jazz07.rchland.ibm.com:21443/jazz/users/param)[in] res The result
1467 *
1468 */
1469 void nfs4_Compound_Free(nfs_res_t *res)
1470 {
1471 release_nfs4_res_compound(res->res_compound4_extended);
1472 }
1473
1474 /**
1475 * [@brief](https://jazz07.rchland.ibm.com:21443/jazz/users/brief) Free a compound data structure
(gdb) p res->res_compound4_extended
$25 = (struct COMPOUND4res_extended *) 0x7f600c061d70
(gdb) p *res->res_compound4_extended
$26 = {res_compound4 = {status = NFS4_OK, tag = {utf8string_len = 0, utf8string_val = 0x0}, resarray = {resarray_len = 3, resarray_val = 0x7f600c063620}}, res_refcnt = 0}
(gdb)

(gdb) f 14
#14 0x00007f6030142a52 in nfs_dupreq_free_dupreq (dv=0x7f600c043ce0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/RPCAL/nfs_dupreq.c:913
913 func->free_function(dv->res);
(gdb) list
908 " cksum %" PRIu64 " %s",
909 dv, dv->hin.tcp.rq_xid, dv->hk,
910 dupreq_state_table[dv->complete]);
911 if (dv->res) {
912 func = nfs_dupreq_func(dv);
913 func->free_function(dv->res);
914 free_nfs_res(dv->res);
915 }
916 PTHREAD_MUTEX_destroy(&dv->dre_mtx);
917 pool_free(dupreq_pool, dv);
(gdb) p *func
$28 = {service_function = 0x7f60301be27b <nfs4_Compound>, free_function = 0x7f60301bedcc <nfs4_Compound_Free>, xdr_decode_func = 0x7f603010704b <xdr_COMPOUND4args>,
  xdr_encode_func = 0x7f60301bf24e <xdr_COMPOUND4res_extended>, funcname = 0x7f6030253f6d "NFS4_COMP", dispatch_behaviour = 4}
(gdb)

/**
* [@brief](https://jazz07.rchland.ibm.com:21443/jazz/users/brief) Deep-free a duplicate request cache entry.
*
* If the entry has processed request data, the corresponding free
* function is called on the result.  The cache entry is then returned
* to the dupreq_pool.
*/
static inline void nfs_dupreq_free_dupreq(dupreq_entry_t *dv)
{
        const nfs_function_desc_t *func;

        assert(dv->refcnt == 0);

        LogDebug(COMPONENT_DUPREQ,
                 "freeing dupreq entry dv=%p, dv xid=%" PRIu32
                 " cksum %" PRIu64 " %s",
                 dv, dv->hin.tcp.rq_xid, dv->hk,
                 dupreq_state_table[dv->complete]);
        if (dv->res) {
                func = nfs_dupreq_func(dv);
                func->free_function(dv->res);
                free_nfs_res(dv->res);
        }
        PTHREAD_MUTEX_destroy(&dv->dre_mtx);
        pool_free(dupreq_pool, dv);
}

(gdb) p *dv->res->res_compound4_extended
$30 = {res_compound4 = {status = NFS4_OK, tag = {utf8string_len = 0, utf8string_val = 0x0}, resarray = {resarray_len = 3, resarray_val = 0x7f600c063620}}, res_refcnt = 0}
(gdb)
(gdb) p dv->hin.tcp.rq_xid
$37 = 3585181026

(gdb) f 15
#15 0x00007f6030142c04 in dupreq_entry_put (dv=0x7f600c043ce0) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/RPCAL/nfs_dupreq.c:946
946 nfs_dupreq_free_dupreq(dv);
(gdb) list
941
942 /* If ref count is zero, no one should be accessing it other
943 * than us.  so no lock is needed.
944 */
945 if (refcnt == 0) {
946 nfs_dupreq_free_dupreq(dv);
947 }
948 }
949
950 /**
(gdb) p (*dv->res->res_compound4_extended).res_compound4.resarray.resarray_val
$39 = (nfs_resop4 *) 0x7f600c063620

/**
* [@brief](https://jazz07.rchland.ibm.com:21443/jazz/users/brief) release a ref count on dupreq_entry_t
*
* The caller must not access dv any more after this call as it could be
* freed here.
*/
static inline void dupreq_entry_put(dupreq_entry_t *dv)
{
        int32_t refcnt;

        refcnt = atomic_dec_uint32_t(&dv->refcnt);

        assert(refcnt >= 0);

        /* If ref count is zero, no one should be accessing it other
         * than us.  so no lock is needed.
         */
        if (refcnt == 0) {
                nfs_dupreq_free_dupreq(dv);
        }
}

(gdb) f 16
#16 0x00007f6030144b34 in nfs_dupreq_finish (reqnfs=0x7f60000383b0, rc=NFS_REQ_OK) at /usr/src/debug/gpfs.nfs-ganesha-5.7-ibm010.00.el8.x86_64/RPCAL/nfs_dupreq.c:1318
1318 dupreq_entry_put(ov);
(gdb) list
1313 ov, ov->hin.tcp.rq_xid,
1314 drc, dupreq_state_table[ov->complete],
1315 ov->refcnt);
1316
1317 /* release hashtable ref count */
1318 dupreq_entry_put(ov);
1319
1320 /* conditionally retire another */
1321 if (cnt++ < DUPREQ_MAX_RETRIES) {
1322 PTHREAD_MUTEX_lock(&drc->drc_mtx);
(gdb)

(gdb) p *dv
$50 = {rbt_k = {left = 0x0, right = 0x0, parent = 0x7f6020018290, red = 1, gen = 10149}, fifo_q = {tqe_next = 0x0, tqe_prev = 0x7f6020006d30}, dupes = {tqh_first = 0x0, tqh_last = 0x7f600001d420}, dre_mtx = {
    __data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 3, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}},
    __size = '\000' <repeats 16 times>, "\003", '\000' <repeats 22 times>, __align = 0}, hin = {addr = {ss_family = 0, __ss_padding = '\000' <repeats 117 times>, __ss_align = 0}, tcp = {rq_xid = 1773307234,
      checksum = 0}, rq_prog = 100003, rq_vers = 4, rq_proc = 1}, hk = 10757007890444590312, complete = true, refcnt = 2, res = 0x7f6000029ff0, rc = NFS_REQ_OK, dupe_cnt = 0}

(gdb) p reqnfs->svc.rq_xprt->xp_u2
$51 = (void *) 0x7f5fec001880

(gdb) p *drc
$52 = {type = DRC_TCP_V4, xt = {npart = 1, flags = 5, cachesz = 127, tree = 0x7f5fec0019c0}, dupreq_q = {tqh_first = 0x7f5ff0013050, tqh_last = 0x7f600001d410}, drc_mtx = {__data = {__lock = 0, __count = 0,
      __owner = 0, __nusers = 0, __kind = 3, __spins = 1, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 16 times>, "\003\000\000\000\001", '\000' <repeats 18 times>,
    __align = 0}, npart = 1, cachesz = 127, size = 64, maxsize = 1024, hiwat = 64, flags = 0, refcnt = 3, retwnd = 0, d_u = {tcp = {addr = {ss_family = 10,
        __ss_padding = "\003l", '\000' <repeats 14 times>, "\377\377\n\000dZ", '\000' <repeats 95 times>, __ss_align = 0}, recycle_k = {left = 0x0, right = 0x0, parent = 0x0, red = 0, gen = 1}, recycle_q = {
        tqe_next = 0x0, tqe_prev = 0x0}, recycle_time = 0, hk = 13144803945354501838}}}

(gdb) p drc->dupreq_q
$53 = {tqh_first = 0x7f5ff0013050, tqh_last = 0x7f600001d410}

(gdb) p ov_hk
$54 = 1318207189023739884

(gdb) p ov
$55 = (dupreq_entry_t *) 0x7f600c043ce0

(gdb) p *ov
$56 = {rbt_k = {left = 0x7f6020089b30, right = 0x7f600c0073c0, parent = 0x7f6020023ec0, red = 1, gen = 0}, fifo_q = {tqe_next = 0x0, tqe_prev = 0x0}, dupes = {tqh_first = 0x0, tqh_last = 0x7f600c043d10},
  dre_mtx = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 3, __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}},
    __size = '\000' <repeats 16 times>, "\003", '\000' <repeats 22 times>, __align = 0}, hin = {addr = {ss_family = 0, __ss_padding = '\000' <repeats 117 times>, __ss_align = 0}, tcp = {rq_xid = 3585181026,
      checksum = 0}, rq_prog = 100003, rq_vers = 4, rq_proc = 1}, hk = 1318207189023739884, complete = true, refcnt = 0, res = 0x7f600c066bb0, rc = NFS_REQ_OK, dupe_cnt = 0}

(gdb) p ov->hin.tcp.rq_xid
$57 = 3585181026

(gdb) p *dv->res->res_compound4_extended
$58 = {res_compound4 = {status = NFS4_OK, tag = {utf8string_len = 0, utf8string_val = 0x0}, resarray = {resarray_len = 1, resarray_val = 0x7f60000485f0}}, res_refcnt = 1}

(gdb) p *ov->res->res_compound4_extended
$59 = {res_compound4 = {status = NFS4_OK, tag = {utf8string_len = 0, utf8string_val = 0x0}, resarray = {resarray_len = 3, resarray_val = 0x7f600c063620}}, res_refcnt = 0}
(gdb)

**Comment by ffilz** [comment]
Can you summarize? Following your gdb session is a lot to absorb

**Comment by ffilz** [comment]
Closing really old issues with no response.
